File: c-sources/devphysiodaq/devphysiodaq_ui/devphysiodaq_dialog.cpp
#include "devphysiodaq_dialog.h"
#include "./ui_devphysiodaq_dialog.h"

devphysiodaq_dialog::devphysiodaq_dialog(QWidget *parent, int n, analogInStruct *analogInfo) : QDialog(parent) , ui(new Ui::devphysiodaq_dialog) {
  
  ui->setupUi(this);

  // fixed size, not resizable
  
  this->setFixedSize(this->width(),this->height());
  setWindowTitle("change channel settings");

  m_numchan = n;
  m_analogPtr = analogInfo;

  // because we make changes to the analog info, a pointer doesn't work when cancel is pressed, 
  // a copy is needed. This copy is returned when OK is pressed, on Cancel it is destroyed

  m_analogInfo = new analogInStruct[m_numchan];
  if (m_analogPtr != nullptr) memcpy(m_analogInfo,m_analogPtr,m_numchan * sizeof(analogInStruct));

  // only numerics are allowed in the gain and offset edit boxes

  ui->gainEdit->setValidator(new QDoubleValidator(0, 100, 2, this));
  ui->offsetEdit->setValidator(new QDoubleValidator(-100, 100, 2, this));
  
  // set the names of the channels to the channel list, make the first channel the default
  // and set the gain and offset in the boxes

  for (int i=0;i<m_numchan;i++) ui->channelComboBox->addItem(m_analogInfo[i].name);

  ui->gainEdit->setText(QString::number(m_analogInfo[0].gain,'f',3));
  ui->offsetEdit->setText(QString::number(m_analogInfo[0].offset,'f',2));
}

devphysiodaq_dialog::~devphysiodaq_dialog() {
  delete ui;
}

// OK is pressed, copy the changed analog info to the parent structure

void devphysiodaq_dialog::on_okButton_clicked() {

  // get the last items from the dialog, (no return can be given)


  int curItem = ui->channelComboBox->currentIndex();
  m_analogInfo[curItem].offset = ui->offsetEdit->text().toDouble();
  m_analogInfo[curItem].gain = ui->gainEdit->text().toDouble();

  memcpy(m_analogPtr,m_analogInfo,m_numchan * sizeof(analogInStruct));

  accept();
}

// cancel is pressed, close 

void devphysiodaq_dialog::on_cancelButton_clicked() {
  
  reject();
}

// offset is changed, store in structure

void devphysiodaq_dialog::on_offsetEdit_editingFinished() {

  int curItem = ui->channelComboBox->currentIndex();
  m_analogInfo[curItem].offset = ui->offsetEdit->text().toDouble();
}

// edit is changed, store in structure

void devphysiodaq_dialog::on_gainEdit_editingFinished() {
  
  int curItem = ui->channelComboBox->currentIndex();
  m_analogInfo[curItem].gain = ui->gainEdit->text().toDouble();
}

// comboxbox is changed, change gain and edit

void devphysiodaq_dialog::on_channelComboBox_currentIndexChanged(int index) {

  int curItem = ui->channelComboBox->currentIndex();
  
  ui->gainEdit->setText(QString::number(m_analogInfo[curItem].gain,'f',4));
  ui->offsetEdit->setText(QString::number(m_analogInfo[curItem].offset,'f',2));
}

File: python/devphysiodaq_dialog.py
#
# devphsyiodaq_dialog
#
#     this class handles the device settings for the physiodaq device. A dialog box is 
#     popped up to let the user change gain and offset from the cnannels


from PySide6.QtWidgets import QDialog
from PySide6.QtGui import QDoubleValidator
from devphysiodaq_ui import Ui_devphysiodaq_dialog

class devPhsyioDaq_dialog(QDialog) :

  # constructor
  #
  #   the UI is loaded

  def __init__(self,analogInfo):

    super().__init__()
    self.ui = Ui_devphysiodaq_dialog()
    self.ui.setupUi(self)

    self.m_analogInfo = analogInfo

    # set the validation for the edit and gain box, only numerics are allowed

    self.ui.gainEdit.setValidator(QDoubleValidator(-5,5,2))
    self.ui.offsetEdit.setValidator(QDoubleValidator(-100,100,2))

    # fill the channel select box with the names of the channels and set the gain and
    # offset in the edit boxes
    
    for analogChannel in self.m_analogInfo :
      self.ui.channelComboBox.addItem(analogChannel["name"])

    self.ui.gainEdit.setText("%6.4f" % self.m_analogInfo[0]["gain"])
    self.ui.offsetEdit.setText("%5.3f" % self.m_analogInfo[0]["offset"])

    # connect to slots

    self.ui.okButton.clicked.connect(self.onOK)
    self.ui.cancelButton.clicked.connect(self.onCancel)
    self.ui.gainEdit.editingFinished.connect(self.onGain)
    self.ui.offsetEdit.editingFinished.connect(self.onOffset)
    self.ui.channelComboBox.currentIndexChanged.connect(self.onSelectChannel)
 
  # OK button
  #
  #   return with accept
         
  def onOK(self) :

    self.accept()

  # cancel button
  #
  #   reject
    
  def onCancel(self) :

    self.reject()

  # gain edit
  #
  #   get the current selected channel and modify the gain of it
    
  def onGain(self) :

    curItem = self.ui.channelComboBox.currentIndex()
    self.m_analogInfo[curItem]["gain"] = float(self.ui.gainEdit.text())

  # offset edit
  #
  #   get the current selected channel and modify the offset of it
    
  def onOffset(self) :

    curItem = self.ui.channelComboBox.currentIndex()
    self.m_analogInfo[curItem]["offset"] = float(self.ui.offsetEdit.text())

  # combobox select channel
  #
  #   retrieve the gain and offset for the selected channel and place this in the 
  #   edit boxes
    
  def onSelectChannel(self) :
    
    curItem = self.ui.channelComboBox.currentIndex()
    self.ui.gainEdit.setText("%6.4f" % self.m_analogInfo[curItem]["gain"])
    self.ui.offsetEdit.setText("%5.2f" % self.m_analogInfo[curItem]["offset"])  

  # getAnalogInfo
  #
  #   returns the (modified) analog structure
    
    def getAnalogInfo(self) :

      return self.m_analogInfo